### DISCUSSION TITLE
How to detect the URL data from a Brave Browser PWA?

### DISCUSSION THREAD

**Discussion by usizu** [johnste/finicky#368] [General]
**Any ideas, suggestions or other feedback**
I've installed a Progressive Web App through Brave Browser to my desktop (email client Proton Mail) and finicky doesn't detect it — even when adding the generalised URL logging from this page: https://github.com/johnste/finicky/wiki/Troubleshooting-&-Tips#i-want-to-see-all-incoming-parameters-to-the-url-handling-functions

The logging works for the same PWA installed from Safari, but not the one from Brave. Nothing shows up at all in the finicky console. 

Is there any way to detect these URLs? Otherwise they always default to opening in Brave Browser, (they are ignored by finicky entirely, which has the default set to Safari).

**Comment by usizu** [comment]
(Given that it works with the PWA generated by Safari but not the one generated by Brave, my reason for using Brave's instead is just that when opened simultaneously, Brave's PWA launches faster)

**Comment by usizu** [comment]
Getting the bundleId with `osascript -e 'id of app "ProtonMail"'` (which is `com.brave.Browser.app.jnpecgipniidlgicjocehkhajgdnjekh`) and then doing

```js
{
  // Open any link clicked in Proton Mail PWA
  match: ({ opener }) =>
    opener.bundleId === "com.brave.Browser.app.jnpecgipniidlgicjocehkhajgdnjekh",
  browser: "Safari"
},
```

also doesn't work

**Comment by ascandroli** [comment]
I found the solution https://github.com/johnste/finicky/issues/153#issuecomment-886952938 
This is what I do for the Google Meet PWA using Brave too.
```
      {
            match: finicky.matchHostnames(["meet.google.com"]),
            browser: ({ urlString }) => ({
                name: "com.brave.Browser",
                profile: "Default",
                args: [
                    '--app-id=kjgfgldnnfoeklkmfkjfagphfepbbdan', // the app ID for the Google Meet application shortcut.
                    `--app-launch-url-for-shortcuts-menu-item=${urlString}`, // notice it is NOT passing 'url' as an array entry, but 'urlString' as string. Chrome Application Shortcuts need a normal string not url objects.

                ],
            }),
        },
```

**Comment by demyxco** [comment]
@usizu did you ever figure this out? I have the same config as you but for Google Chrome. Nothing works.

**Comment by usizu** [comment]
> @usizu did you ever figure this out? I have the same config as you but for Google Chrome. Nothing works.

Never figured it out unfortunately, I currently just use PWAs from Safari ironically 😅

**Comment by noamsiegel** [comment]
Damn, I hope that this is fixed.

I currently use Vivaldi PWAs and am having the same issue. I am trying to open Google Meet URLs in my Vivaldi PWA, but it just opens the PWA without actually redirecting to the Google Meet URL itself.

@demyxco @usizu Let me know if you ever get a solution working! I know that v2 is in the works, so fingers crossed. I will contribute to it when it is published.

**Comment by demyxco** [comment]
@noamsiegel I ended up using Browserosaurus for now.

**Comment by usizu** [comment]
> I found the solution [#153 (comment)](https://github.com/johnste/finicky/issues/153#issuecomment-886952938) This is what I do for the Google Meet PWA using Brave too.
> 
> ```
>       {
>             match: finicky.matchHostnames(["meet.google.com"]),
>             browser: ({ urlString }) => ({
>                 name: "com.brave.Browser",
>                 profile: "Default",
>                 args: [
>                     '--app-id=kjgfgldnnfoeklkmfkjfagphfepbbdan', // the app ID for the Google Meet application shortcut.
>                     `--app-launch-url-for-shortcuts-menu-item=${urlString}`, // notice it is NOT passing 'url' as an array entry, but 'urlString' as string. Chrome Application Shortcuts need a normal string not url objects.
> 
>                 ],
>             }),
>         },
> ```

I'm trying to understand this. It looks like it only works if the destination URL is always consistent?

Also is `--app-id=kjgfgldnnfoeklkmfkjfagphfepbbdan` the bundleId of the PWA?

**Comment by ascandroli** [comment]
Hi @usizu, I'm sorry, but I'm not entirely sure if I understood your question correctly. Generally, it works for any meet.google.com link. They are always the same. But I'm really not sure if that is the "consistency" you were referring to.

2) Yes! It is the last portion of the bundleID, in your example it would be: `jnpecgipniidlgicjocehkhajgdnjekh` which is also the internal app-id that Chrome based browsers use to identify apps and extensions.

**Comment by usizu** [comment]
I was referring to the destination URL - it seems your solution only works for links that go to Google Meet. It wouldn’t work for example for any arbitrary link with random destination that I would click in a PWA email client, I think?

Whereas you want any link that triggers a Google Meeting to open in the Meeting PWA client

**Comment by ascandroli** [comment]
Oh, I see! I misunderstood earlier, I'm sorry for the confusion. Now I understand why nothing is showing up in the Finicky logs.

Chrome-based browsers handle links with the HTTP protocol internally, keeping them within the app instead of sending them to the default browser. Here are a few options you could try, each with its pros and cons:

1. Use a bookmarklet to redirect the page once it’s opened in the PWA:
`javascript:var url=window.location.href;var op='?%27;if(window.location.search){op='%27&%27;}url=url+op+%27finicky=brave%27;url=url.replace(window.location.protocol,%27finicky:%27);window.location.href=url;`

2. Use an extension like “Open with Switchbar.” It’s similar to the Finicky extension which I believe is only available for Firefox.
3. Create a TamperMonkey script to rewrite all links on a page to use the finicky:// protocol. Clicking these links will open Finicky first.
4. Use an extension to automatically redirect all new tabs or windows to Finicky. I use (and admittedly abuse) an extension called LeechBlock. I have a whitelist of pages allowed in Brave and automatically redirect the rest using code similar to the bookmarklet.	
⚠️ This method can easily lead to infinite loops and has caused me more than a few headaches.
